File: Python/JDI/web/selenium/elements/complex/table/cell.py
from JDI.web.selenium.driver.utils.web_driver_by_utils import WebDriverByUtils
from JDI.web.selenium.elements.api_interact.find_element_by import By
from JDI.web.selenium.elements.base.select_element import SelectElement


class Cell(SelectElement):
    row_index = None
    column_index = None
    table = None
    column_num = None
    row_nom = None
    web_element = None
    column_name = None
    row_name = None
    cell_locator_template = By.xpath(".//tr[{1}]/td[{0}]")

    def __init__(self, column_num, row_num, col_name, row_name, table,
                 cell_locator_template=None, web_element=None, column_index=None, row_index=None):
        if web_element is not None:
            self.web_element = web_element
        if cell_locator_template is not None:
            self.cell_locator_template = cell_locator_template
        if column_index is not None:
            self.column_index = column_index + 1 if table.rows.has_header and table.rows.by_line_template is None else column_index
        if row_index is not None:
            self.row_index = row_index

        self.column_num = column_num
        self.row_num = row_num
        self.col_name = col_name
        self.row_name = row_name
        self.table = table

    def set_web_element(self, web_element):
        self.web_element = web_element

    def update_data(self, col_name, row_name):
        if self.column_name is None or self.column_name == "" and not (col_name is None or col_name == ""):
            self.column_name = col_name
        if self.row_name is None and self.row_name == "" and not (row_name is None or row_name == ""):
            self.row_name = row_name
        return self

    def get_text_action(self):
        return self.get().get_text()

    def get(self):
        cell = SelectElement(web_element=self.web_element) if self.web_element is not None else \
            SelectElement(
                WebDriverByUtils.fill_by_template(self.cell_locator_template, [self.column_index, self.row_index]))
        cell.init(parent=self.table, avatar=cell.avatar)
        return cell

File: Java/JDI/jdi-uitest-win/src/main/java/com/epam/jdi/uitests/win/winnium/elements/complex/table/Cell.java
package com.epam.jdi.uitests.win.winnium.elements.complex.table;

import com.epam.jdi.uitests.core.interfaces.MapInterfaceToElement;
import com.epam.jdi.uitests.core.interfaces.base.IBaseElement;
import com.epam.jdi.uitests.core.interfaces.complex.tables.interfaces.ICell;
import com.epam.jdi.uitests.win.winnium.elements.base.Element;
import com.epam.jdi.uitests.win.winnium.elements.base.SelectElement;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;
import sun.reflect.generics.reflectiveObjects.NotImplementedException;

import static com.epam.commons.LinqUtils.last;
import static com.epam.jdi.uitests.core.settings.JDISettings.exception;
import static com.epam.jdi.uitests.win.winnium.driver.WebDriverByUtils.fillByMsgTemplate;

public class Cell extends Element implements ICell {
    private int rowIndex;
    private int columnIndex;
    private Table table;
    private int columnNum;
    private int rowNum;
    private String columnName;
    private String rowName;
    private By cellLocatorTemplate = By.xpath(".//tr[{1}]/td[{0}]");

    Cell(WebElement webElement, int columnNum, int rowNum, String colName, String rowName,
         By cellLocatorTemplate, Table table) {
        this.getAvatar().setWebElement(webElement);
        this.columnNum = columnNum;
        this.rowNum = rowNum;
        this.columnName = colName;
        this.rowName = rowName;
        if (cellLocatorTemplate != null)
            this.cellLocatorTemplate = cellLocatorTemplate;
        this.table = table;
    }

    Cell(int columnIndex, int rowIndex, int columnNum, int rowNum, String colName, String rowName,
         By cellLocatorTemplate, Table table) {
        this.columnIndex = (((Rows)table.rows()).hasHeader && ((Rows)table.rows()).lineTemplate == null) ? columnIndex + 1 : columnIndex;
        this.rowIndex = rowIndex;
        this.columnNum = columnNum;
        this.rowNum = rowNum;
        this.columnName = colName;
        this.rowName = rowName;
        if (cellLocatorTemplate != null)
            this.cellLocatorTemplate = cellLocatorTemplate;
        this.table = table;
    }

    public void setWebElement(WebElement webElement) {
        this.getAvatar().setWebElement(webElement);
    }

    public int columnNum() {
        return columnNum;
    }

    public int rowNum() {
        return rowNum;
    }

    public String columnName() {
        return (columnName != null && !columnName.equals(""))
                ? columnName
                : table.columns().headers().get(columnNum - 1);
    }

    public String rowName() {
        return (rowName != null && !rowName.equals(""))
                ? rowName
                : table.rows().headers().get(rowNum - 1);
    }

    protected String getTextAction() {
        return get().getText();
    }

    protected void clickAction() {
        get().click();
    }

    protected boolean isSelectedAction() {
        return get().isSelected();
    }

    private SelectElement get() {
        SelectElement cell = avatar.hasWebElement()
                ? new SelectElement(getWebElement())
                : new SelectElement(getDriver().findElement(fillByMsgTemplate(cellLocatorTemplate, columnIndex, rowIndex)));
        //TODO
        //cell.init(table, cell.getAvatar());
        return cell;
    }

    public WebElement get(By subLocator) {
        //TODO
        return null;
        //return get().get(subLocator);
    }

    public <T extends IBaseElement> T get(Class<T> clazz) {
        T instance;
        try {
            instance = (clazz.isInterface())
                    ? (T) MapInterfaceToElement.getClassFromInterface(clazz).newInstance()
                    : clazz.newInstance();
            Element el = ((Element)instance);
            el.setAvatar(getAvatar());
            el.setLocator(fillByMsgTemplate(cellLocatorTemplate, columnIndex, rowIndex));
            //TODO
            //el.init(table, instance.getAvatar());
        } catch (Exception ex) {
            throw exception("Can't get Cell from interface/class: " + last((clazz + "").split("\\.")));
        }
        return get(instance);
    }

    public <T extends IBaseElement> T get(T cell) {
        return null;
        //TODO
        /*
        BaseElement cellSelect = (BaseElement) cell;
        if (cellSelect.getAvatar().hasWebElement())
            return cell;
        By locator = cellSelect.getLocator();
        if (locator == null || locator.toString().equals(""))
            locator = cellLocatorTemplate;
        if (!locator.toString().contains("{0}") || !locator.toString().contains("{1}"))
            throw exception("Can't create cell with locator template " + cellSelect.getLocator()
                    + ". Template for Cell should contains '{0}' - for column and '{1}' - for row indexes.");
        cellSelect.avatar.setDriverName(avatar.getDriverName());
        cellSelect.init(table,
                new GetElementModule(
                        fillByMsgTemplate(locator, columnIndex, rowIndex), cellSelect));
        return cell;*/
    }

    public Cell updateData(String colName, String rowName) {
        if ((columnName == null || columnName.equals("")) && !(colName == null || colName.equals("")))
            columnName = colName;
        if ((this.rowName == null || this.rowName.equals("")) && !(rowName == null || rowName.equals("")))
            this.rowName = rowName;
        return this;
    }

    public String getValue() {
        throw new NotImplementedException();
    }

    public void click() {
        throw new NotImplementedException();

    }

    public void select() {
        throw new NotImplementedException();

    }

    public boolean isSelected() {
        throw new NotImplementedException();
    }

    public String getText() {
        throw new NotImplementedException();
    }

    public String waitText(String text) {
        throw new NotImplementedException();
    }

    public String waitMatchText(String regEx) {
        throw new NotImplementedException();
    }
}
